File: YoonJunSik/week_3/키패드누르기.java
class Solution {
    public String solution(int[] numbers, String hand) {
        StringBuilder answer = new StringBuilder();

        // 키패드 좌표 매핑
        int[][] keypad = {
                {3,1}, // 0
                {0,0}, // 1
                {0,1}, // 2
                {0,2}, // 3
                {1,0}, // 4
                {1,1}, // 5
                {1,2}, // 6
                {2,0}, // 7
                {2,1}, // 8
                {2,2}  // 9
        };

        // 시작 위치
        int[] leftPos = {3, 0};  // *
        int[] rightPos = {3, 2}; // #

        for (int num : numbers) {
            // 왼쪽 열
            if (num == 1 || num == 4 || num == 7) {
                answer.append("L");
                leftPos = keypad[num];
            }
            // 오른쪽 열
            else if (num == 3 || num == 6 || num == 9) {
                answer.append("R");
                rightPos = keypad[num];
            }
            // 가운데 열
            else {
                int leftDist = Math.abs(leftPos[0] - keypad[num][0]) + Math.abs(leftPos[1] - keypad[num][1]);
                int rightDist = Math.abs(rightPos[0] - keypad[num][0]) + Math.abs(rightPos[1] - keypad[num][1]);

                if (leftDist < rightDist) {
                    answer.append("L");
                    leftPos = keypad[num];
                } else if (rightDist < leftDist) {
                    answer.append("R");
                    rightPos = keypad[num];
                } else {
                    if (hand.equals("right")) {
                        answer.append("R");
                        rightPos = keypad[num];
                    } else {
                        answer.append("L");
                        leftPos = keypad[num];
                    }
                }
            }
        }
        return answer.toString();
    }
}

File: ParkJuHyun/week_3/키패드누르기.py
def solution(numbers, hand):
    answer = ''
    
    pad = {'1':(0,0), '2':(0,1), '3':(0,2),
           '4':(1,0), '5':(1,1), '6':(1,2),
           '7':(2,0), '8':(2,1), '9':(2,2),
           '*':(3,0), '0':(3,1), '#':(3,2)
          }
    
    left = pad['*'] 
    right = pad['#']
    
    for num in numbers:
        if num == 1 or num == 4 or num == 7 :
            answer += 'L'      
            left = pad[str(num)]   
        
        elif num == 3 or num == 6 or num == 9 :
            answer += 'R'      
            right = pad[str(num)]
        else:
            # 1. 타겟 번호의 좌표
            target_pos = pad[str(num)]
            
            # 2. 현재 왼손/오른손에서 타겟까지의 맨해튼 거리 계산
            left_dist = abs(left[0] - target_pos[0]) + abs(left[1] - target_pos[1])
            right_dist = abs(right[0] - target_pos[0]) + abs(right[1] - target_pos[1])
            
            # 3. 거리 비교
            if left_dist < right_dist:
                answer += 'L'
                left = target_pos
            elif left_dist > right_dist:
                answer += 'R'
                right = target_pos
            else:
                if hand == 'left':
                    answer += 'L'
                    left = target_pos
                else:
                    answer += 'R'
                    right = target_pos
            
    return answer